### PR TITLE
Add comma-separated multi-term search (#10 Phase 3)

### LINKED ISSUE #10: Improve search query performance and add multi-term search support
## Summary

This issue tracks a set of improvements to the `posts_clauses()` search query in `public/class-media-search-enhanced.php`, motivated by a [WordPress.org support request](https://wordpress.org/support/topic/search-for-multiple-images-at-once-in-the-wordpress-media-library/) asking for multi-image search capability, and a review of the existing query structure.

### 1. Replace LEFT JOINs + DISTINCT with EXISTS subqueries (performance)

**Problem:** The current query LEFT JOINs `postmeta` and taxonomy tables (`term_relationships`, `term_taxonomy`, `terms`), which multiplies result rows — one attachment with N meta rows produces N result rows. `DISTINCT` is then used to deduplicate, which is expensive on large result sets.

**Solution:** Replace JOINs with `EXISTS` subqueries:

```sql
-- Current (slow): JOIN then DISTINCT
LEFT JOIN postmeta AS mse_pm ON posts.ID = mse_pm.post_id
WHERE ... OR (mse_pm.meta_key = '_wp_attachment_image_alt' AND mse_pm.meta_value LIKE '%term%')
DISTINCT

-- Improved (faster): EXISTS, no DISTINCT needed
WHERE ... OR EXISTS (
  SELECT 1 FROM postmeta
  WHERE post_id = posts.ID
  AND meta_key = '_wp_attachment_image_alt'
  AND meta_value LIKE '%term%'
)
```

`EXISTS` short-circuits on the first match and eliminates row multiplication entirely. No more `DISTINCT` needed.

**Impact:** Biggest performance gain, especially for large media libraries (10k+ attachments).

### 2. Fix ID search to use integer comparison (performance)

**Problem:** Currently `posts.ID LIKE '%term%'` forces a string cast on an integer primary key column, preventing index usage.

**Solution:** Only check ID when the search term is numeric, using `posts.ID = %d` instead of LIKE.

**Impact:** Small but free improvement — allows the primary key index to be used when searching by ID.

### 3. Add multi-term (comma-separated) search support (feature)

**Problem:** Users cannot search for multiple files at once. Searching `"image-a.jpg, photo-2.jpg"` looks for the literal string instead of matching either file. See [support request](https://wordpress.org/support/topic/search-for-multiple-images-at-once-in-the-wordpress-media-library/).

**Solution:**
- Detect commas in the search string
- Split by comma and trim each term; discard empty terms
- Generate the full set of search conditions for each term
- OR the per-term groups together (match if **any** term matches)
- Cap at 10 terms max to prevent abuse / excessive query size
- Single-term searches (no commas) behave identically to current behavior

**Impact:** Quality-of-life improvement for content editors managing large media libraries who need to find/bulk-delete specific files.

## Implementation Plan

Implement incrementally in this order:

1. **Phase 1:** EXISTS subqueries + remove DISTINCT (performance foundation)
2. **Phase 2:** Fix ID comparison to use integer match when numeric
3. **Phase 3:** Add multi-term search with comma delimiter and 10-term cap

All changes are confined to the `posts_clauses()` method in `public/class-media-search-enhanced.php`.

## Backward Compatibility

- Single-term searches produce identical results
- No database changes required
- No UI changes required
- No new dependencies

### PR DESCRIPTION
**Pull request by yoren**
## Summary

- Users can now search for multiple terms separated by commas (e.g. `image-a.jpg, photo-2.jpg`)
- Each term is matched across all fields (title, alt text, filename, GUID, content, excerpt, taxonomy)
- Results include attachments matching **any** of the terms (OR logic)
- Capped at 10 terms to prevent abuse / excessive query size
- Single-term searches (no commas) produce identical SQL to before

Addresses the [WordPress.org support request](https://wordpress.org/support/topic/search-for-multiple-images-at-once-in-the-wordpress-media-library/) for multi-image search.

## What changed

**`public/class-media-search-enhanced.php`**:
- Extracted `build_search_conditions()` private static method — generates the full set of SQL conditions (ID, posts fields, postmeta EXISTS, taxonomy EXISTS) for a single term
- `posts_clauses()` now splits the search string by commas, trims/filters/caps at 10, calls `build_search_conditions()` per term, and ORs the groups together
- Taxonomy filter is built once and shared across all terms

**`tests/SearchTest.php`** — 5 new correctness tests:
| Test | What it verifies |
|---|---|
| `test_comma_separated_search` | `"alpha, bravo"` finds both, not a non-match |
| `test_single_term_without_comma_unchanged` | No behavioral change for single terms |
| `test_comma_search_across_fields` | One term matches by title, another by alt text |
| `test_comma_search_ignores_empty_terms` | `",, term ,,"` works correctly |
| `test_comma_search_caps_at_10_terms` | 11th and 12th terms are dropped |

**`tests/benchmark/QueryStructureTest.php`**:
- New `test_multi_term_search_generates_or_groups` — asserts both terms appear in SQL with >= 4 EXISTS subqueries (2 per term)
- Updated `capture_query()` to match on first comma-separated term

## Test plan

- [x] All 32 tests pass (22 correctness + 9 benchmark structural + 1 EXPLAIN log)
- [x] Single-term search behavior unchanged (existing tests pass without modification)
- [x] Multi-term: each term generates its own full search condition group
- [x] Empty terms from extra commas are ignored
- [x] 10-term cap enforced
- [ ] CI passes across PHP 7.4–8.3

Closes #10.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/1fixdotio/media-search-enhanced/pull/15" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: public/class-media-search-enhanced.php [hunk 1]
```diff
@@ -189,61 +189,93 @@ public static function posts_clauses( $pieces ) {
...
-
-			// Use exact integer match for ID when search term is numeric; skip ID match otherwise.
-			$search_trimmed = trim( $vars['s'] );
-			$search_int     = absint( $search_trimmed );
-			if ( $search_int > 0 && ctype_digit( $search_trimmed ) ) {
-				$id_condition = sprintf( "($wpdb->posts.ID = %d)", $search_int );
-			} else {
-				$id_condition = '(1=0)';
+			// Split search into comma-separated terms; cap at 10 to prevent abuse.
+			$terms = array_values( array_filter( array_map( 'trim', explode( ',', $vars['s'] ) ) ) );
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: Good catch. Fixed in 8c29b2d — added `'strlen'` callback to `array_filter()` so `'0'` is preserved as a valid search term.

File: public/class-media-search-enhanced.php [hunk 2]
```diff
@@ -189,61 +189,93 @@ public static function posts_clauses( $pieces ) {
...
+			if ( empty( $terms ) ) {
+				$terms = array( $vars['s'] );
 			}
+			$terms = array_slice( $terms, 0, 10 );
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: Correct — this is the intentional feature change from the [support request](https://wordpress.org/support/topic/search-for-multiple-images-at-once-in-the-wordpress-media-library/). Will document in the changelog. Literal comma search is an extremely unlikely use case for media library search compared to the multi-file lookup this enables.

File: public/class-media-search-enhanced.php [hunk 3]
```diff
@@ -189,61 +189,93 @@ public static function posts_clauses( $pieces ) {
...
+	private static function build_search_conditions( $term, $taxes, $tax_filter ) {
+		global $wpdb;
+
+		$like = '%' . $wpdb->esc_like( $term ) . '%';
+
+		// Use exact integer match for ID when search term is numeric.
+		$search_trimmed = trim( $term );
+		$search_int     = absint( $search_trimmed );
+		if ( $search_int > 0 && ctype_digit( $search_trimmed ) ) {
+			$id_condition = sprintf( "($wpdb->posts.ID = %d)", $search_int );
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php:278]: This is the same `sprintf()` + `absint()` pattern that was reviewed and accepted in PR #14 — Mesa actually suggested switching from `$wpdb->prepare()` to `sprintf()` in that PR to avoid nested prepare calls. Using `$wpdb->prepare('%s.ID = %d')` would quote `$wpdb->posts` as a string literal, producing invalid SQL. `sprintf` with `%d` + `absint()` is the correct approach here.

File: public/class-media-search-enhanced.php [hunk 4]
```diff
@@ -189,61 +189,93 @@ public static function posts_clauses( $pieces ) {
...
-					. $wpdb->prepare( "t.slug LIKE %s OR tt.description LIKE %s OR t.name LIKE %s", $like, $like, $like )
-					. "))";
+			// Generate search conditions for each term, then OR them together.
+			$term_groups = array();
+			foreach ( $terms as $term ) {
+				$term_groups[] = self::build_search_conditions( $term, $taxes, $tax_filter );
 			}
 
-			$pieces['where'] .= " )";
+			$pieces['where'] .= " AND ( " . implode( ' OR ', $term_groups ) . " )";
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: The 10-term cap exists specifically to bound query complexity. With EXISTS subqueries (Phase 1), each term group short-circuits independently — MySQL does not scan the full table for each OR. The profiling tools (`composer test:profile`) are available for users to validate performance at their scale.

File: public/class-media-search-enhanced.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: Good catch. Fixed in 8c29b2d — now returns `$pieces` unmodified (no search rewrite) when all terms are empty.

File: public/class-media-search-enhanced.php [hunk 5]
```diff
@@ -189,61 +189,93 @@ public static function posts_clauses( $pieces ) {
...
-			$search_int     = absint( $search_trimmed );
-			if ( $search_int > 0 && ctype_digit( $search_trimmed ) ) {
-				$id_condition = sprintf( "($wpdb->posts.ID = %d)", $search_int );
-			} else {
-				$id_condition = '(1=0)';
+			// Split search into comma-separated terms; cap at 10 to prevent abuse.
+			$terms = array_values( array_filter( array_map( 'trim', explode( ',', $vars['s'] ) ), 'strlen' ) );
+			if ( empty( $terms ) ) {
+				return $pieces;
 			}
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: Great catch — this was a real data exposure bug. Fixed in 3a25cfb.

The early return now appends `AND 1=0` before returning, so the query returns zero results instead of all attachments. Added `test_all_commas_search_returns_empty` to cover this edge case.

File: tests/SearchTest.php [hunk 6]
```diff
@@ -391,4 +386,117 @@ public function test_request_query_ajax_fallback() {
...
+	/**
+	 * Test 21c: All-commas search returns no results (not all attachments).
+	 */
+	public function test_all_commas_search_returns_empty() {
+		$id = $this->create_attachment( array( 'post_title' => 'should-not-appear' ) );
+
+		$results = $this->search_attachments( ',,,' );
+
+		$this->assertEmpty( $results, 'Search with only commas should return no results.' );
+	}
```
[the remark this answers is not in the document]
Comment (yoren) [reply tests/SearchTest.php:548]: Good catch — the test was passing by coincidence (unauthenticated path searched for literal commas). Fixed in 4c55b79: now authenticates as admin so it actually exercises the comma-splitting → empty-terms → `AND 1=0` guard.

File: public/class-media-search-enhanced.php [hunk 7]
```diff
@@ -189,61 +193,102 @@ public static function posts_clauses( $pieces ) {
...
+				$terms = array_values( array_filter( array_map( 'trim', explode( ',', $vars['s'] ) ), 'strlen' ) );
 			} else {
-				$id_condition = '(1=0)';
+				$terms = array( $vars['s'] );
 			}
+			if ( empty( $terms ) ) {
+				$pieces['where'] .= " AND 1=0";
+				return $pieces;
+			}
+			$terms = array_slice( $terms, 0, max( 1, $max_terms ) );
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: Good catch — fixed in e293f32. Now requires `is_admin() && DOING_AJAX` (both set by `admin-ajax.php`) in addition to `action=query-attachments`. Appending `?action=query-attachments` to a frontend URL no longer triggers comma splitting.

Also added `mse_is_media_modal_request` filter for testability and edge cases.

File: public/class-media-search-enhanced.php [hunk 8]
```diff
@@ -144,7 +144,11 @@ public static function posts_clauses( $pieces ) {
 
 		// Rewrite the where clause
 		if ( ! empty( $vars['s'] ) && ( ( isset( $_REQUEST['action'] ) && 'query-attachments' == $_REQUEST['action'] ) || 'attachment' == $vars['post_type'] ) ) {
-			$pieces['where'] = " AND $wpdb->posts.post_type = 'attachment' AND ($wpdb->posts.post_status = 'inherit' OR $wpdb->posts.post_status = 'private')";
+			$status_clause = "$wpdb->posts.post_status = 'inherit'";
+			if ( current_user_can( 'read_private_posts' ) ) {
+				$status_clause .= " OR $wpdb->posts.post_status = 'private'";
+			}
+			$pieces['where'] = " AND $wpdb->posts.post_type = 'attachment' AND ($status_clause)";
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: Noted — this is pre-existing behavior from the original plugin (v0.6.0). Worth addressing in a separate issue to avoid scope creep in this PR.

File: public/class-media-search-enhanced.php [hunk 9]
```diff
@@ -127,7 +128,48 @@ public function load_plugin_textdomain() {
...
+	 * @param string   $search Search SQL fragment.
+	 * @param WP_Query $query  The WP_Query instance.
+	 * @return string
+	 */
+	public static function suppress_core_search( $search, $query ) {
+		if ( self::is_mse_search() ) {
+			return '';
+		}
+		return $search;
+	}
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php:161]: Good catch — fixed in 93423e2. Now checks `$query->query_vars` directly instead of the global, so secondary queries (sidebar widgets, etc.) won't have their search suppressed.

File: public/class-media-search-enhanced.php [hunk 10]
```diff
@@ -142,106 +184,123 @@ public static function posts_clauses( $pieces ) {
...
-				$year = substr( $vars['m'], 0, 4 );
-				$monthnum = substr( $vars['m'], 4 );
-				$pieces['where'] .= $wpdb->prepare( " AND YEAR($wpdb->posts.post_date) = %d AND MONTH($wpdb->posts.post_date) = %d", $year, $monthnum );
-			} else {
-				if ( ! empty( $vars['year'] ) && 'false' != $vars['year'] ) {
-					$pieces['where'] .= $wpdb->prepare( " AND YEAR($wpdb->posts.post_date) = %d", $vars['year'] );
-				}
+		if ( ! self::is_mse_search() ) {
+			return $pieces;
+		}
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php]: Agreed this is a pre-existing architectural issue — `posts_clauses` has used global `$wp_query` since v0.6.0. Fixing it properly requires changing the filter registration to accept 2 args and updating all the query_vars reads. Worth a separate issue to avoid scope creep.

File: public/class-media-search-enhanced.php [hunk 11]
```diff
@@ -127,7 +128,51 @@ public function load_plugin_textdomain() {
...
+	 */
+	public static function suppress_core_search( $search, $query ) {
+		$vars = $query->query_vars;
+		if ( ! empty( $vars['s'] )
+			&& ( ( isset( $_REQUEST['action'] ) && 'query-attachments' == $_REQUEST['action'] )
+				|| ( isset( $vars['post_type'] ) && 'attachment' == $vars['post_type'] ) ) ) {
+			return '';
+		}
+		return $search;
+	}
```
[the remark this answers is not in the document]
Comment (yoren) [reply public/class-media-search-enhanced.php:161]: Fixed in b23a164. Both `suppress_core_search` and `posts_clauses` now use the same `$query` instance:

- `posts_clauses($pieces, $query)` — registration changed to accept 2 args
- `is_mse_search($query)` — takes the WP_Query parameter instead of reading global
- No more global `$wp_query` dependency in any of these methods

Secondary WP_Query instances are now handled correctly — both filters evaluate the same query object.
